lianad/src/payjoin/receiver.rs

```diff
@@ -0,0 +1,358 @@
+
+    persister.update_metadata(
+        Some(PayjoinStatus::Signing),
+        Some(psbt.unsigned_tx.compute_txid()),
+        Some(psbt.clone()),
+        Some(bip21.clone()),
+    );
+
+    Ok(())
+}
+
+fn finalize_proposal(
+    proposal: Receiver<ProvisionalProposal>,
+    persister: &ReceiverPersister,
+    history: SessionHistory,
+    db_conn: &mut Box<dyn DatabaseConnection>,
+    secp: &secp256k1::Secp256k1<secp256k1::VerifyOnly>,
+) -> Result<(), Box<dyn Error>> {
+    if let Some(proposed_psbt) = history.psbt_with_contributed_inputs() {
+        let txid = proposed_psbt.unsigned_tx.compute_txid();
+        if let Some(psbt) = db_conn.spend_tx(&txid) {
+            let mut is_signed = false;
+            for psbtin in &psbt.inputs {
+                if !psbtin.partial_sigs.is_empty() {
+                    log::debug!("[Payjoin] PSBT is signed!");
+                    is_signed = true;
+                    break;
+                }
+            }
+
+            if is_signed {
+                let mut psbt = psbt.clone();
+                finalize_psbt(&mut psbt, secp);
+
+                let proposal = proposal
+                    .finalize_proposal(
+                        |_| Ok(psbt.clone()),
+                        None,
+                        Some(FeeRate::from_sat_per_vb(150).unwrap()),
+                    )
+                    .save(persister)?;
+
+                send_payjoin_proposal(proposal, persister, history)?;
+            }
+        }
+    }
+    Ok(())
+}
+
+fn send_payjoin_proposal(
+    mut proposal: Receiver<PayjoinProposal>,
+    persister: &ReceiverPersister,
+    history: SessionHistory,
+) -> Result<(), Box<dyn Error>> {
+    let (req, ctx) = proposal
+        .extract_req(OHTTP_RELAY)
+        .expect("Failed to extract request");
+
+    let psbt = proposal.psbt().clone();
+    let txid = psbt.unsigned_tx.compute_txid();
+
+    // Respond to sender
+    log::info!("[Payjoin] Receiver responding to sender...");
+    match post_request(req) {
+        Ok(resp) => {
+            proposal
+                .process_res(resp.bytes().expect("Failed to read response").as_ref(), ctx)
+                .save(persister)?;
+
+            let bip21 = history.pj_uri().unwrap();
+            persister.update_metadata(
+                Some(PayjoinStatus::Completed),
+                Some(txid),
+                Some(psbt),
+                Some(bip21.to_string()),
+            );
+        }
+        Err(err) => log::error!("[Payjoin] send_payjoin_proposal(): {}", err),
+    }
+    Ok(())
+}
+
+fn process_receiver_session(
+    db: &sync::Arc<sync::Mutex<dyn DatabaseInterface>>,
+    bit: &mut sync::Arc<sync::Mutex<dyn BitcoinInterface>>,
+    desc: &descriptors::LianaDescriptor,
+    secp: &secp256k1::Secp256k1<secp256k1::VerifyOnly>,
+) -> Result<(), Box<dyn Error>> {
+    let mut db_conn = db.connection();
+    for (session_id, session) in db_conn.payjoin_get_all_receiver_sessions() {
```

Comment:
This db call should filter out sessions that are completed by using the completed_at timestamp being set. That way we dont need to check the metadata.